### DISCUSSION TITLE
[vim9script] There are advantages on having `import` keyword defaulting to `import` dir?

### DISCUSSION THREAD

**Discussion by saccarosium** [vim/vim#16107] [General]
Hi,
I was wondering if there are any benefits of making `import` default to the `import` directory.

Right now I don't see any benefit in using it over `import autoload` and I believe that it is a poor design choice for this reason:
1. `import`, for my understanding, will load all the file at once increasing startup time
2. having _yet another_ directory for plugins can make things confusing (especially for new users and courageous vimscript learners)
3. `import`, for my understanding, automatically create `function#names` calls for legacy script

So my idea would be to make `import` defaulting to the `autoload` directory instead

**Comment by bfrg** [comment]
Rule of thumb:
- Use the `autoload/` directory for functions that are supposed to be lazy-loaded. So basically anything that stands on the RHS of a mapping, command or autocmd. My script files in the `plugin/` directory only use `import autoload "../foo.vim"`. (The relative path helps Vim find the correct file faster. There's an issue about this somewhere on github which I can't find right now.)
- Use the `import/` directory for sharing functionality between script files. It's a like a lib directory. If a function (or class, variable, etc.) is used in several scripts, you can put it in there and export it. Other script files can `import` it. So a script file in `autoload/` would import another script file using `import` (and not `import autoload`).

For smaller plugins you probably won't need the `import/` directory but if you write larger plugins, it's quite helpful. And see also `:h vim9-import`.

> import, for my understanding, will load all the file at once increasing startup time

Script files are always loaded "all at once".

**Reply by saccarosium** [reply]
> Script files are always loaded "all at once".

I've explained purely. What I meant is that if I do `import <path>` on a `plugin/` file won't lazy load, impacting startup time.

> Use the import/ directory for sharing functionality between script files. It's a like a lib directory.

Yes, I understand that. But why wouldn't you just use the `autoload/` directory for that anyway?

Let me try to explain what I mean with two example:

> [!NOTE]
> Lines that starts with:
> - `Q:` means that it is directed to the reader (e.g. Q: Why would I prefer X over Y?)
> - `A:` it is a possible reasoning I the writer came up with
> - `ReplyToA:` my reason why the assumption **I** made up isn't valid or not optimal 

> [!NOTE]
> I will use `"` for comments so it easier to read but I know that `#` is the correct syntax for comments.

## Example 1

INTENT: Writing a plugin for terminal management (something like [this](https://github.com/hahdookin/miniterm.vim/blob/main/autoload/miniterm.vim))

1. I want to make a class that represent a terminal instance.
2. I know I want to reuse this class in other files. I can put this either in `autoload/mypluginname/terminalclass.vim` or in `import/mypluginname/terminalclass.vim`.

Q1: Why would I choose the `import/mypluginname/terminalclass` over `autoload/mypluginname/terminalclass.vim`?
Q2: Are there any down side/performance issues of using one over the other?
Q3: There is a better solution that doesn't introduce a all new directory in the `runtimepath`?
Q4: Since the majority of the time the plugin author/user **wants** the lazy loading (basing this on the plugin I list in [awesome-vim9](https://github.com/saccarosium/awesome-vim9) where there are some very complicated plugins that won't ever use the `import` directory). Isn't better to flip the behaviour so import **is** lazyloaded and looks into `autoload/` and having something like `import noautoload` when I simply want to source the all script without lazy loading (which is rare).

## Example 2

INTENT: Writing a library plugin that other plugins can use (something like [this](https://github.com/lifepillar/vim-devel))

1. I want to create a utility path class for others to use.
2. I want to put in a place where it is easily usable by other plugin authors and wont impact startup time.

Basically the same question I used above.
